### PR TITLE
US-506187 - Enhance Cassandra client helm configuration options

### PR DESCRIPTION
**Pull request by baylp2022**
US-506187 - Enhance Cassandra client helm configuration options

### CODE REVIEW

**Comment by Saurabh-16** [comment]
Please fix the go test cases.

File: charts/pega/README.md [hunk 1]
```diff
@@ -672,6 +672,15 @@ Parameter     | Description | Default value
 `trustStorePassword` | If required provide trustStorePassword value in plain text. For better security leave this parameter blank and include the password in an external secrets manager with the key CASSANDRA_TRUSTSTORE_PASSWORD. | Empty
 `keyStore` | If required, provide the keystore certificate file name.<br/>When using a keystore certificate, you must also include a Kubernetes secret name that contains the keystore certificate in the global.certificatesSecrets parameter.<br/>Pega deployments only support keystores that use the Java Key Store (.jks) format. | Empty
 `keyStorePassword` | If required provide keyStorePassword value in plain text. For better security leave this parameter blank and include the password in an external secrets manager with the key CASSANDRA_KEYSTORE_PASSWORD. | Empty
+`asyncProcessingEnabled*` | Specify a prefix to use when creating Pega-managed keyspaces in Cassandra. | false
```
Comment (taz-pega-work) [inline charts/pega/README.md:675]: is this correct? I see "Specify a prefix to use when creating Pega-managed keyspaces in Cassandra." is also used to describe keyspacesPrefix. maybe copy error?

File: charts/pega/README.md [hunk 2]
```diff
@@ -672,6 +672,15 @@ Parameter     | Description | Default value
...
+`keyspacesPrefix` | Specify a prefix to use when creating Pega-managed keyspaces in Cassandra. |
```
Comment (taz-pega-work) [inline charts/pega/README.md:676]: needs a basic blank following the | for that column

Is it ""?? or Empty?

File: charts/pega/README.md [hunk 3]
```diff
@@ -672,6 +672,15 @@ Parameter     | Description | Default value
...
+`extendedTokenAwarePolicy*` | Enable an extended token aware policy for use when a Cassandra range query runs. When enabled this policy selects a token from the token range to determine which Cassandra node to send the request. Before you can enable this policy, you must configure the token range partitioner. | false
+`latencyAwarePolicy*` | Enable a latency awareness policy, which collects the latencies of the queries for each Cassandra node and maintains a per-node latency score (an average). | false
+`customRetryPolicy*` | Enable the use of a customized retry policy for your Pega Platform deployment. After enabling this policy in your deployment configuration, Cassandra queries that timeout will be retried. The number of retries may be configured using the dynamic system setting (DSS): dnode/cassandra_custom_retry_policy/retryCount. If not configured, queries will be retried once. | false
```
Comment (taz-pega-work) [inline charts/pega/README.md:679]: avoid passive voice by using active voice>>
Enable the use of a customized retry policy for your Pega Platform deployment. After you enable this policy in your deployment configuration, the deployment retries Cassandra queries that timeout. Configure the number of retries using the dynamic system setting (DSS): dnode/cassandra_custom_retry_policy/retryCount. The default is 1, so if you do not specify a retry count, timed out queries are retried once.

File: charts/pega/README.md [hunk 4]
```diff
@@ -672,6 +672,15 @@ Parameter     | Description | Default value
...
+`speculativeExecutionPolicy*` | Enable the speculative execution policy for retrieving data from your Cassandra service. When enabled, the Pega Platform will send a query to multiple nodes in your Cassandra service and process the first query response. This provides lower perceived latencies at the client, but puts greater load on your Cassandra service. | false
```
Comment (taz-pega-work) [inline charts/pega/README.md:680]: Enable the speculative execution policy for retrieving data from your Cassandra service. When enabled, the Pega Platform sends a query to multiple nodes in your Cassandra service and processes the first query response. This provides lower perceived latencies at the client (level? layer?), but puts a greater load on your Cassandra service.

File: charts/pega/templates/pega-environment-config.yaml [hunk 5]
```diff
@@ -59,6 +59,38 @@ data:
   CASSANDRA_TRUSTSTORE: "{{ .Values.dds.trustStore }}"
   # Path to a mounted key store file for Cassandra encryption
   CASSANDRA_KEYSTORE: "{{ .Values.dds.keyStore }}"
+  # Enable asynchronous processing of records in DDS Dataset save operation. Failures to store individual records will
+  # not interrupt Dataset save operations
```
Comment (taz-pega-work) [inline charts/pega/templates/pega-environment-config.yaml:63]: consistency...missing period>>not interrupt Dataset save operations.
;-(

File: charts/pega/templates/pega-environment-config.yaml [hunk 6]
```diff
@@ -59,6 +59,38 @@ data:
...
+  CASSANDRA_ASYNC_PROCESSING_ENABLED: "{{ .Values.dds.asyncProcessingEnabled }}"
+  # Specify a prefix to use when creating Pega-managed keyspaces in Cassandra
```
Comment (taz-pega-work) [inline charts/pega/templates/pega-environment-config.yaml:65]: consistency...missing period>>Specify a prefix to use when creating Pega-managed keyspaces in Cassandra.

File: charts/pega/templates/pega-environment-config.yaml [hunk 7]
```diff
@@ -59,6 +59,38 @@ data:
...
+  # token from the token range to determine which Cassandra node to send the request. Before you can enable this policy,
+  # you must configure the token range partitioner.
+  CASSANDRA_EXTENDED_TOKEN_AWARE_POLICY: "{{ .Values.dds.extendedTokenAwarePolicy }}"
+  # Enable a latency awareness policy, which collects the latencies of the queries for each Cassandra node and maintains
+  # a per-node latency score (an average).
+  CASSANDRA_LATENCY_AWARE_POLICY: "{{ .Values.dds.latencyAwarePolicy }}"
+  # Enable the use of a customized retry policy for your Pega Platform deployment. After enabling this policy in your
+  # deployment configuration, Cassandra queries that timeout will be retried. The number of retries may be configured
+  # using the dynamic system setting (DSS): dnode/cassandra_custom_retry_policy/retryCount. If not configured, queries
+  # will be retried once
```
Comment (taz-pega-work) [inline charts/pega/templates/pega-environment-config.yaml:77]: consistency...missing period>>are retried once.

File: charts/pega/templates/pega-environment-config.yaml [hunk 8]
```diff
@@ -59,6 +59,38 @@ data:
...
+  CASSANDRA_CUSTOM_RETRY_POLICY: "{{ .Values.dds.customRetryPolicy }}"
+  # Enable the speculative execution policy for retrieving data from your Cassandra service. When enabled, the Pega
+  # Platform will send a query to multiple nodes in your Cassandra service and process the first query response. This
+  # provides lower perceived latencies at the client, but puts greater load on your Cassandra service.
+  CASSANDRA_SPECULATIVE_EXECUTION_POLICY: "{{ .Values.dds.speculativeExecutionPolicy }}"
+  # Enable reporting of DDS SDK metrics to a Java Management Extension (JMX) format for use by your organization to
+  # monitor your Cassandra service. Setting this property `false` disables metrics being exposed through the JMX
+  # interface; disabling also limits the metrics being collected using the DDS landing page
```
Comment (taz-pega-work) [inline charts/pega/templates/pega-environment-config.yaml:85]: consistency...missing period>>interface; disabling also limits the metrics being collected using the DDS landing page.

File: charts/pega/templates/pega-environment-config.yaml [hunk 9]
```diff
@@ -59,6 +59,38 @@ data:
...
+  CASSANDRA_JMX_METRICS_ENABLED: "{{ .Values.dds.jmxMetricsEnabled }}"
+  # Enable reporting of DDS SDK metrics to a Comma Separated Value (CSV) format for use by your organization to monitor
+  # your Cassandra service. If you enable this property, use the Pega Platform DSS:
+  # dnode/ddsclient/metrics/csv_directory to customize the filepath to which the deployment writes CSV files. By
+  # default, after you enable this property, CSV files will be written to the Pega Platform work directory
```
Comment (taz-pega-work) [inline charts/pega/templates/pega-environment-config.yaml:90]: consistency...missing period>>CSV files are written to the Pega Platform work directory.

File: charts/pega/values.yaml [hunk 10]
```diff
@@ -347,6 +347,38 @@ dds:
   # username,password,trustStorePassword,keyStorePassword parameters in External Secrets Manager, and enter the external secret for the credentials
   # make sure the keys in the secret should be CASSANDRA_USERNAME, CASSANDRA_PASSWORD , CASSANDRA_TRUSTSTORE_PASSWORD , CASSANDRA_KEYSTORE_PASSWORD
   external_secret_name: ""
+  # Enable asynchronous processing of records in DDS Dataset save operation. Failures to store individual records will
+  # not interrupt Dataset save operations
```
Comment (taz-pega-work) [inline charts/pega/values.yaml:351]: same simple comments apply in this chart as they did in the above chart.

**Review by taz-pega-work — CHANGES_REQUESTED** [review]
getting there. thanks for asking for a review!!

**Review by taz-pega-work — APPROVED** [review]
This is a great enhancement! thank you. Could you add an intro describing what the example delivers so it's clear what parameters like the trustStore and keyStore, policy, and metrics signify in the Cassandra service? I think it'd be super handy to users.

thank you for going out of your way to clarify all of these improvements.

File: charts/pega/README.md (hunk 1 above)
Comment (baylp2022) [reply charts/pega/README.md:675]: Done

File: charts/pega/README.md (hunk 2 above)
Comment (baylp2022) [reply charts/pega/README.md:676]: Fixed - changed to Empty

File: charts/pega/README.md (hunk 3 above)
Comment (baylp2022) [reply charts/pega/README.md:679]: Fixed

File: charts/pega/README.md (hunk 4 above)
Comment (baylp2022) [reply charts/pega/README.md:680]: Fixed: changed 'at the client' to 'for your deployment'

File: charts/pega/templates/pega-environment-config.yaml (hunk 6 above)
Comment (baylp2022) [reply charts/pega/templates/pega-environment-config.yaml:65]: Fixed

File: charts/pega/templates/pega-environment-config.yaml (hunk 7 above)
Comment (baylp2022) [reply charts/pega/templates/pega-environment-config.yaml:77]: Fixed

File: charts/pega/templates/pega-environment-config.yaml (hunk 8 above)
Comment (baylp2022) [reply charts/pega/templates/pega-environment-config.yaml:85]: Fixed

File: charts/pega/templates/pega-environment-config.yaml (hunk 9 above)
Comment (baylp2022) [reply charts/pega/templates/pega-environment-config.yaml:90]: Fixed

**Review by taz-pega-work — APPROVED** [review]
Thanks Paul. These changes are ace. Clients will be grateful!! ;-)

**Pull request closed without merging**